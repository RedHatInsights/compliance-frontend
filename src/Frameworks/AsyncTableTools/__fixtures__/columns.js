import { severityLevels } from './items';

export default [
  {
    title: 'ID',
    sortByProperty: 'id',
    exportKey: 'id',
    key: 'id',
    manageable: false,
  },
  {
    title: 'Name',
    key: 'id',
    sortByProperty: 'name',
    exportKey: 'name',
    renderFunc: (_a, _b, item) => item.name,
  },
  {
    title: 'Another Name column',
    key: 'another_name_column',
    sortByProperty: 'name',
    exportKey: 'name',
    renderExport: (name) => `${name} via export render`,
  },
  {
    title: 'Description',
    key: 'desc',
    sortByProperty: 'description',
    renderExport: (item) => `${item.name} description rendered for export`,
  },
  {
    title: 'Severity sorted by function',
    sortByFunction: (item) => item.name,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Severity by Array',
    sortByProperty: 'severity',
    sortByArray: severityLevels,
    exportKey: 'severity',
    renderFunc: (_a, _b, item) => item.severity,
  },
  {
    title: 'Download button',
    key: 'download',
    manageable: false,
    export: false,
    renderFunc: () => 'download link',
  },
];
