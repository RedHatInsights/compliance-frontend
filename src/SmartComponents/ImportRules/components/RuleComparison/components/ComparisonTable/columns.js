import React from 'react';
import { nowrap } from '@patternfly/react-table';

import RuleCell from './components/RuleCell';
import ProfileVersionCell from './components/ProfileVersionCell';

const rule = {
  title: <strong>Rule</strong>,
  Component: RuleCell,
};

export const profileVersion = (version, targetVersion, additionalProps) => ({
  dataLabel: `RHEL ${version.os_major_version}.${version.os_minor_version} -  SSG version ${version.ssg_version}`,
  title: (
    <>
      <strong>
        RHEL {version.os_major_version}.{version.os_minor_version}
      </strong>
      <br />
      SSG version {version.ssg_version}
    </>
  ),
  Component: (props) => (
    <ProfileVersionCell
      {...{
        version,
        targetVersion,
        ...props,
        ...additionalProps,
      }}
    />
  ),
  transforms: [nowrap],
});

export default [rule];
