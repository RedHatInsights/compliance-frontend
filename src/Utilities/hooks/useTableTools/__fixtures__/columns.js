import { severityLevels } from './items';

export default [
    {
        title: 'Name',
        sortByProperty: 'name',
        renderFunc: (_a, _b, item) => (
            item.name
        )
    }, {
        title: 'Description',
        sortByProperty: 'description'
    }, {
        title: 'Severity sorted by function',
        sortByFunction: (item) => (item.name),
        renderFunc: (_a, _b, item) => (
            item.severity
        )
    }, {
        title: 'Severity by Array',
        sortByProperty: 'severity',
        sortByArray: severityLevels,
        renderFunc: (_a, _b, item) => (
            item.severity
        )
    }
];
