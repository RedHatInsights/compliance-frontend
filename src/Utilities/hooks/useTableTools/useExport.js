import { camelCase, getProperty } from 'Utilities/helpers';

const CSV_FILE_PREFIX = 'compliance-export';
const CSV_DELIMITER = ',';
const ENCODINGS = {
  csv: 'text/csv',
  json: 'application/json',
};

const filename = (format) =>
  CSV_FILE_PREFIX + '-' + new Date().toISOString() + '.' + format;

const encoding = (format) => `data:${ENCODINGS[format]};charset=utf-8`;

export const linkAndDownload = (data, filename) => {
  const link = document.createElement('a');
  link.href = data;
  link.download = filename;
  link.click();
};

const textForCell = (row, column) => {
  const { exportKey, renderExport } = column;
  let cell = exportKey ? getProperty(row, exportKey) : row;
  if (renderExport) {
    return renderExport(cell);
  } else {
    return cell;
  }
};

export const csvForItems = ({ items, columns }) => {
  const header = columns
    .map((column) => column.original || column.title)
    .join(CSV_DELIMITER);
  const csvRows = [
    header,
    ...items.map((row) =>
      columns
        .map((column) => `"${textForCell(row, column)}"`)
        .join(CSV_DELIMITER)
    ),
  ];

  return encodeURI(`${encoding('csv')},${csvRows.join('\n')}`);
};

export const jsonForItems = ({ items, columns }) => {
  const result = items.map((row) =>
    columns.reduce((object, column) => {
      const key = camelCase(column.original || column.title);
      const value = textForCell(row, column);

      object[key] = value;
      return object;
    }, {})
  );

  return encodeURI(`${encoding('json')},${JSON.stringify(result)}`);
};

const callCallback = (callback, ...args) => callback && callback(...args);

const useExport = ({
  exporter,
  columns = [],
  isDisabled = false,
  onStart,
  onComplete,
  onError,
}) => {
  const exportableColumns = columns.filter(
    (column) =>
      column.export !== false && (column.exportKey || column.renderExport)
  );
  const exportWithFormat = async (format) => {
    callCallback(onStart);
    const items = await exporter()
      .then((items) => {
        callCallback(onComplete, items);
        return items;
      })
      .catch((error) => callCallback(onError, error));

    const formater = format === 'csv' ? csvForItems : jsonForItems;

    if (items) {
      return linkAndDownload(
        formater({
          items,
          columns: exportableColumns,
        }),
        filename(format)
      );
    } else {
      console.info('No items returned for export');
      return;
    }
  };

  return {
    toolbarProps: {
      exportConfig: {
        isDisabled,
        onSelect: (_, format) => exportWithFormat(format),
      },
    },
  };
};

export const useExportWithItems = (items, columns, options = {}) => {
  const exportEnabled = options?.exportable;
  const {
    columns: exportableColumns,
    onStart,
    onComplete,
  } = typeof options.exportable === 'object' ? options.exportable : {};
  const exportableSelectedColumns = (exportableColumns || columns).filter(
    (column) => columns.includes(column)
  );

  const exportProps = useExport({
    exporter: () => Promise.resolve(items),
    columns: exportableSelectedColumns,
    isDisabled: items.length === 0,
    onStart,
    onComplete,
  });

  return exportEnabled ? exportProps : {};
};

export default useExport;
