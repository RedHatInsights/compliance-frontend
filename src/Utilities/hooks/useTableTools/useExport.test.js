import { act, renderHook } from '@testing-library/react';
import columns from './__fixtures__/columns';
import items from './__fixtures__/items';
import useExport, { jsonForItems, csvForItems } from './useExport';

const exampleItems = items(25);

describe('useExport', () => {
  const exporter = jest.fn(() => Promise.resolve(exampleItems));
  const defaultOptions = {
    exporter,
    columns,
  };

  it('returns an export config toolbar config', () => {
    const { result } = renderHook(() => useExport(defaultOptions));
    expect(result.current.toolbarProps.exportConfig).toBeDefined();
    expect(result).toMatchSnapshot();
  });

  it('returns an export config toolbar config', () => {
    const { result } = renderHook(() =>
      useExport({
        ...defaultOptions,
        isDisabled: true,
      })
    );
    expect(result.current.toolbarProps.exportConfig.isDisabled).toBe(true);
  });

  it('calls the exporter via onSelect', () => {
    const { result } = renderHook(() => useExport(defaultOptions));

    act(() => {
      result.current.toolbarProps.exportConfig.onSelect(null, 'csv');
    });

    expect(exporter).toHaveBeenCalled();

    act(() => {
      result.current.toolbarProps.exportConfig.onSelect(null, 'json');
    });

    expect(exporter).toHaveBeenCalled();
  });
});

describe('jsonForItems', () => {
  it('returns an json export of items', () => {
    const result = jsonForItems({
      columns,
      items: exampleItems,
    });

    expect(result).toMatchSnapshot();
  });
});

describe('csvForItems', () => {
  it('returns an csv export of items', () => {
    const result = csvForItems({
      columns,
      items: exampleItems,
    });

    expect(result).toMatchSnapshot();
  });
});
