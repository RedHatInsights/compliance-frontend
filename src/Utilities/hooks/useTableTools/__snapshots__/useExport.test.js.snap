// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csvForItems returns an csv export of items 1`] = `"data:text/csv;charset=utf-8,ID,Name,Another%20Name%20column,Description,Severity%20sorted%20by%20function,Severity%20by%20Array%0A13,TEST%20ITEM%20#13,TEST%20ITEM%20#13%20via%20export%20render,TEST%20ITEM%20#13%20description%20rendered%20for%20export,medium,medium%0A1,TEST%20ITEM%20#1,TEST%20ITEM%20#1%20via%20export%20render,TEST%20ITEM%20#1%20description%20rendered%20for%20export,medium,medium%0A3,TEST%20ITEM%20#3,TEST%20ITEM%20#3%20via%20export%20render,TEST%20ITEM%20#3%20description%20rendered%20for%20export,unknown,unknown%0A4,TEST%20ITEM%20#4,TEST%20ITEM%20#4%20via%20export%20render,TEST%20ITEM%20#4%20description%20rendered%20for%20export,high,high%0A5,TEST%20ITEM%20#5,TEST%20ITEM%20#5%20via%20export%20render,TEST%20ITEM%20#5%20description%20rendered%20for%20export,medium,medium%0A6,TEST%20ITEM%20#6,TEST%20ITEM%20#6%20via%20export%20render,TEST%20ITEM%20#6%20description%20rendered%20for%20export,low,low%0A7,TEST%20ITEM%20#7,TEST%20ITEM%20#7%20via%20export%20render,TEST%20ITEM%20#7%20description%20rendered%20for%20export,unknown,unknown%0A8,TEST%20ITEM%20#8,TEST%20ITEM%20#8%20via%20export%20render,TEST%20ITEM%20#8%20description%20rendered%20for%20export,high,high%0A9,TEST%20ITEM%20#9,TEST%20ITEM%20#9%20via%20export%20render,TEST%20ITEM%20#9%20description%20rendered%20for%20export,medium,medium%0A10,TEST%20ITEM%20#10,TEST%20ITEM%20#10%20via%20export%20render,TEST%20ITEM%20#10%20description%20rendered%20for%20export,low,low%0A11,TEST%20ITEM%20#11,TEST%20ITEM%20#11%20via%20export%20render,TEST%20ITEM%20#11%20description%20rendered%20for%20export,unknown,unknown%0A12,TEST%20ITEM%20#12,TEST%20ITEM%20#12%20via%20export%20render,TEST%20ITEM%20#12%20description%20rendered%20for%20export,high,high%0A2,TEST%20ITEM%20#2,TEST%20ITEM%20#2%20via%20export%20render,TEST%20ITEM%20#2%20description%20rendered%20for%20export,low,low%0A14,TEST%20ITEM%20#14,TEST%20ITEM%20#14%20via%20export%20render,TEST%20ITEM%20#14%20description%20rendered%20for%20export,low,low%0A15,TEST%20ITEM%20#15,TEST%20ITEM%20#15%20via%20export%20render,TEST%20ITEM%20#15%20description%20rendered%20for%20export,unknown,unknown%0A16,TEST%20ITEM%20#16,TEST%20ITEM%20#16%20via%20export%20render,TEST%20ITEM%20#16%20description%20rendered%20for%20export,high,high%0A17,TEST%20ITEM%20#17,TEST%20ITEM%20#17%20via%20export%20render,TEST%20ITEM%20#17%20description%20rendered%20for%20export,medium,medium%0A18,TEST%20ITEM%20#18,TEST%20ITEM%20#18%20via%20export%20render,TEST%20ITEM%20#18%20description%20rendered%20for%20export,low,low%0A19,TEST%20ITEM%20#19,TEST%20ITEM%20#19%20via%20export%20render,TEST%20ITEM%20#19%20description%20rendered%20for%20export,unknown,unknown%0A20,TEST%20ITEM%20#20,TEST%20ITEM%20#20%20via%20export%20render,TEST%20ITEM%20#20%20description%20rendered%20for%20export,high,high%0A21,TEST%20ITEM%20#21,TEST%20ITEM%20#21%20via%20export%20render,TEST%20ITEM%20#21%20description%20rendered%20for%20export,medium,medium%0A22,TEST%20ITEM%20#22,TEST%20ITEM%20#22%20via%20export%20render,TEST%20ITEM%20#22%20description%20rendered%20for%20export,low,low%0A23,TEST%20ITEM%20#23,TEST%20ITEM%20#23%20via%20export%20render,TEST%20ITEM%20#23%20description%20rendered%20for%20export,unknown,unknown%0A24,TEST%20ITEM%20#24,TEST%20ITEM%20#24%20via%20export%20render,TEST%20ITEM%20#24%20description%20rendered%20for%20export,high,high%0A25,TEST%20ITEM%20#25,TEST%20ITEM%20#25%20via%20export%20render,TEST%20ITEM%20#25%20description%20rendered%20for%20export,medium,medium"`;

exports[`jsonForItems returns an json export of items 1`] = `"data:application/json;charset=utf-8,%5B%7B%22id%22:13,%22name%22:%22TEST%20ITEM%20#13%22,%22another_name_column%22:%22TEST%20ITEM%20#13%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#13%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D,%7B%22id%22:1,%22name%22:%22TEST%20ITEM%20#1%22,%22another_name_column%22:%22TEST%20ITEM%20#1%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#1%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D,%7B%22id%22:3,%22name%22:%22TEST%20ITEM%20#3%22,%22another_name_column%22:%22TEST%20ITEM%20#3%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#3%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22unknown%22,%22severity_by_array%22:%22unknown%22%7D,%7B%22id%22:4,%22name%22:%22TEST%20ITEM%20#4%22,%22another_name_column%22:%22TEST%20ITEM%20#4%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#4%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22high%22,%22severity_by_array%22:%22high%22%7D,%7B%22id%22:5,%22name%22:%22TEST%20ITEM%20#5%22,%22another_name_column%22:%22TEST%20ITEM%20#5%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#5%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D,%7B%22id%22:6,%22name%22:%22TEST%20ITEM%20#6%22,%22another_name_column%22:%22TEST%20ITEM%20#6%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#6%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22low%22,%22severity_by_array%22:%22low%22%7D,%7B%22id%22:7,%22name%22:%22TEST%20ITEM%20#7%22,%22another_name_column%22:%22TEST%20ITEM%20#7%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#7%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22unknown%22,%22severity_by_array%22:%22unknown%22%7D,%7B%22id%22:8,%22name%22:%22TEST%20ITEM%20#8%22,%22another_name_column%22:%22TEST%20ITEM%20#8%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#8%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22high%22,%22severity_by_array%22:%22high%22%7D,%7B%22id%22:9,%22name%22:%22TEST%20ITEM%20#9%22,%22another_name_column%22:%22TEST%20ITEM%20#9%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#9%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D,%7B%22id%22:10,%22name%22:%22TEST%20ITEM%20#10%22,%22another_name_column%22:%22TEST%20ITEM%20#10%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#10%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22low%22,%22severity_by_array%22:%22low%22%7D,%7B%22id%22:11,%22name%22:%22TEST%20ITEM%20#11%22,%22another_name_column%22:%22TEST%20ITEM%20#11%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#11%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22unknown%22,%22severity_by_array%22:%22unknown%22%7D,%7B%22id%22:12,%22name%22:%22TEST%20ITEM%20#12%22,%22another_name_column%22:%22TEST%20ITEM%20#12%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#12%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22high%22,%22severity_by_array%22:%22high%22%7D,%7B%22id%22:2,%22name%22:%22TEST%20ITEM%20#2%22,%22another_name_column%22:%22TEST%20ITEM%20#2%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#2%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22low%22,%22severity_by_array%22:%22low%22%7D,%7B%22id%22:14,%22name%22:%22TEST%20ITEM%20#14%22,%22another_name_column%22:%22TEST%20ITEM%20#14%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#14%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22low%22,%22severity_by_array%22:%22low%22%7D,%7B%22id%22:15,%22name%22:%22TEST%20ITEM%20#15%22,%22another_name_column%22:%22TEST%20ITEM%20#15%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#15%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22unknown%22,%22severity_by_array%22:%22unknown%22%7D,%7B%22id%22:16,%22name%22:%22TEST%20ITEM%20#16%22,%22another_name_column%22:%22TEST%20ITEM%20#16%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#16%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22high%22,%22severity_by_array%22:%22high%22%7D,%7B%22id%22:17,%22name%22:%22TEST%20ITEM%20#17%22,%22another_name_column%22:%22TEST%20ITEM%20#17%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#17%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D,%7B%22id%22:18,%22name%22:%22TEST%20ITEM%20#18%22,%22another_name_column%22:%22TEST%20ITEM%20#18%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#18%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22low%22,%22severity_by_array%22:%22low%22%7D,%7B%22id%22:19,%22name%22:%22TEST%20ITEM%20#19%22,%22another_name_column%22:%22TEST%20ITEM%20#19%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#19%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22unknown%22,%22severity_by_array%22:%22unknown%22%7D,%7B%22id%22:20,%22name%22:%22TEST%20ITEM%20#20%22,%22another_name_column%22:%22TEST%20ITEM%20#20%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#20%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22high%22,%22severity_by_array%22:%22high%22%7D,%7B%22id%22:21,%22name%22:%22TEST%20ITEM%20#21%22,%22another_name_column%22:%22TEST%20ITEM%20#21%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#21%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D,%7B%22id%22:22,%22name%22:%22TEST%20ITEM%20#22%22,%22another_name_column%22:%22TEST%20ITEM%20#22%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#22%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22low%22,%22severity_by_array%22:%22low%22%7D,%7B%22id%22:23,%22name%22:%22TEST%20ITEM%20#23%22,%22another_name_column%22:%22TEST%20ITEM%20#23%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#23%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22unknown%22,%22severity_by_array%22:%22unknown%22%7D,%7B%22id%22:24,%22name%22:%22TEST%20ITEM%20#24%22,%22another_name_column%22:%22TEST%20ITEM%20#24%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#24%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22high%22,%22severity_by_array%22:%22high%22%7D,%7B%22id%22:25,%22name%22:%22TEST%20ITEM%20#25%22,%22another_name_column%22:%22TEST%20ITEM%20#25%20via%20export%20render%22,%22description%22:%22TEST%20ITEM%20#25%20description%20rendered%20for%20export%22,%22severity_sorted_by_function%22:%22medium%22,%22severity_by_array%22:%22medium%22%7D%5D"`;

exports[`useExport returns an export config toolbar config 1`] = `
Object {
  "all": Array [
    Object {
      "toolbarProps": Object {
        "exportConfig": Object {
          "isDisabled": false,
          "onSelect": [Function],
        },
      },
    },
  ],
  "current": Object {
    "toolbarProps": Object {
      "exportConfig": Object {
        "isDisabled": false,
        "onSelect": [Function],
      },
    },
  },
  "error": undefined,
}
`;
